h/api/nipsa/logic.py

```diff
@@ -0,0 +1,51 @@
+import json
+
+from h.api.nipsa import models
+
+
+def _publish(request, data):
+    request.get_queue_writer().publish("nipsa_user_requests", data)
+
+
+def index():
+    """Return a list of all the NIPSA'd user IDs.
+
+    :rtype: list of unicode strings
+
+    """
+    return [nipsa_user.userid for nipsa_user in models.NipsaUser.all()]
+
+
+def nipsa(request, userid):
+    """NIPSA a user.
+
+    Add the given user's ID to the list of NIPSA'd user IDs.
+    If the user is already NIPSA'd then nothing will happen (but a "nipsa"
+    message for the user will still be published to the queue).
+
+    """
+    nipsa_user = models.NipsaUser.get_by_id(userid)
+    if not nipsa_user:
+        nipsa_user = models.NipsaUser(userid)
+        request.db.add(nipsa_user)
+
+    _publish(request, json.dumps({"action": "nipsa", "userid": userid}))
+
+
+def unnipsa(request, userid):
+    """Un-NIPSA a user.
+
+    If the user isn't NIPSA'd then nothing will happen (but an "unnipsa"
```

Comment:
1. It's simpler to just always publish, doesn't do any harm and there's no need to increase the complexity of the logic and add more tests

2. This way enables re-unnipsa'ing users. For example if I unnipsa a user but the worker process crashes and doesn't get all of their annotations, I just unnipsa them again to try again. This is now a bit of a mute point since the actual HTML UI for nipsa will make you re-nipsa them in order to unnipsa them again anyway! But originally I had a JSON API (and we may want one again one day) and you could re-nipsa and re-unnipsa directly.